src/main/host/syscall/format.rs

```diff
@@ -0,0 +1,520 @@
+        simple_pointer_impl!($($types),+);
+    };
+    ($type:ty) => {
+        impl SyscallPtrDisplay for SyscallPtr<*const $type> {
+            fn fmt(
+                &self,
+                f: &mut std::fmt::Formatter<'_>,
+                mem: &MemoryManager,
+            ) -> std::fmt::Result {
+                match mem.memory_ref(TypedPluginPtr::new::<$type>(self.ptr, 1)) {
+                    Ok(vals) => write!(f, "{} ({:p})", &(*vals)[0], self.ptr),
+                    // if we couldn't read the memory, just show the pointer instead
+                    Err(_) => write!(f, "{:p}", self.ptr),
+                }
+            }
+        }
+    };
+}
+
+/// Display the pointer without dereferencing any plugin memory. Useful for types like *void, or
+/// when we don't care about the data.
+macro_rules! safe_pointer_impl {
+    ($type:ty, $($types:ty),+) => {
+        safe_pointer_impl!($type);
+        safe_pointer_impl!($($types),+);
+    };
+    ($type:ty) => {
+        impl SyscallPtrDisplay for SyscallPtr<*const $type> {
+            fn fmt(
+                &self,
+                f: &mut std::fmt::Formatter<'_>,
+                _mem: &MemoryManager,
+            ) -> std::fmt::Result {
+                write!(f, "{:p}", self.ptr)
+            }
+        }
+    };
+}
+
+/// Display the array pointer and data. Accesses plugin memory. Can only be used for pod types.
+macro_rules! simple_array_impl {
+    ($type:ty, $($types:ty),+) => {
+        simple_array_impl!($type);
+        simple_array_impl!($($types),+);
+    };
+    ($type:ty) => {
+        impl<const K: usize> SyscallPtrDisplay for SyscallPtr<[$type; K]> {
+            fn fmt(
+                &self,
+                f: &mut std::fmt::Formatter<'_>,
+                mem: &MemoryManager,
+            ) -> std::fmt::Result {
+                match mem.memory_ref(TypedPluginPtr::new::<$type>(self.ptr, K)) {
+                    Ok(vals) => write!(f, "{:?} ({:p})", &(*vals), self.ptr),
+                    // if we couldn't read the memory, just show the pointer instead
+                    Err(_) => write!(f, "{:p}", self.ptr),
+                }
+            }
+        }
+    };
+}
+
+// implement conversions from `SysCallReg`
+
+impl TryFromSyscallReg for nix::fcntl::OFlag {
+    fn try_from_reg(reg: SysCallReg) -> Option<Self> {
+        Self::from_bits(reg.into())
+    }
+}
+
+impl TryFromSyscallReg for nix::sys::eventfd::EfdFlags {
+    fn try_from_reg(reg: SysCallReg) -> Option<Self> {
+        Self::from_bits(reg.into())
+    }
+}
+
+impl TryFromSyscallReg for nix::sys::socket::AddressFamily {
+    fn try_from_reg(reg: SysCallReg) -> Option<Self> {
+        Self::from_i32(reg.into())
+    }
+}
+
+impl TryFromSyscallReg for nix::sys::socket::MsgFlags {
+    fn try_from_reg(reg: SysCallReg) -> Option<Self> {
+        Self::from_bits(reg.into())
+    }
+}
+
+// implement display formatting
+
+simple_display_impl!(i8, i16, i32, i64, isize);
```

Comment:
I think we could reduce the macro usage here while keeping the boilerplate low with an extra layer of Traits. IMO this would make the code easier to read  and debug, though maybe I just need to get more comfortable with Rust macros :).

I think it'd go something like:

```
trait SimpleSyscallDataDisplay : Display {};

impl SyscallDataDisplay for SimpleSyscallDataDisplay {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self)
    }
};

impl SimpleSyscallDataDisplay for i8;
impl SimpleSyscallDataDisplay for i16;
impl SimpleSyscallDataDisplay for i32;
...
```

Reply:
I'll merge now, and then if there's a better way to do it I can make a new PR later.